pennylane/compiler/compiler.py

```diff
@@ -0,0 +1,134 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Compiler developer functions"""
+
+from typing import List
+from importlib import reload
+from collections import defaultdict
+import dataclasses
+import pkg_resources
+
+
+@dataclasses.dataclass
+class AvailableCompilers:
+    """This contains data of installed PennyLane compiler packages"""
+
+    names_entrypoints = {}
+
+
+def _refresh_compilers():
+    """Scan installed PennyLane compiler pacakges to refresh the compilers
+    names and entry points
+    """
+
+    reload(pkg_resources)
+
+    # Refresh the list of compilers
+    AvailableCompilers.names_entrypoints = defaultdict(dict)
+
+    # Iterator packages entry-points with the 'pennylane.compilers' group name
+    for entry in pkg_resources.iter_entry_points("pennylane.compilers"):
+        # Only need name of the parent module
+        module_name = entry.module_name.split(".")[0]
+        AvailableCompilers.names_entrypoints[module_name][entry.name] = entry
+
+
+def available_compilers() -> List[str]:
+    """Loads and returns a list of available compilers that are
+    installed and compatible with the :func:`~.qjit` decorator.
+
+    **Example**
+
+    This method returns the name of installed compiler packages supported in
+    PennyLane. For example, after installing the `Catalyst <https://github.com/pennylaneai/catalyst>`__
+    compiler, this will now appear as an available compiler:
+
+    >>> qml.compiler.available_compilers()
+    ['catalyst']
+    """
+
+    # This class updates the class variable names_entrypoints
+    _refresh_compilers()
+
+    return list(AvailableCompilers.names_entrypoints.keys())
+
+
+def available(compiler="catalyst") -> bool:
+    """Check the availability of the given compiler package.
+
+    Args:
+        compiler (str): name of the compiler package (default value is ``catalyst``)
+
+    Return:
+        bool: ``True`` if the compiler package is installed on the system
+
+    **Example**
+
+    Before installing the ``pennylane-catalyst`` package:
+
+    >>> qml.compiler.available("catalyst")
+    False
+
+    After installing the ``pennylane-catalyst`` package:
+
+    >>> qml.compiler.available("catalyst")
+    True
+    """
+
+    # It only refreshes the compilers names and entry points if the name
+    # is not already stored. This reduces the number of re-importing
+    # ``pkg_resources`` as it can be a very slow operation on systems
+    # with a large number of installed packages.
+
+    if compiler not in AvailableCompilers.names_entrypoints:
+        # This class updates the class variable names_entrypoints
+        _refresh_compilers()
+
+    return compiler in AvailableCompilers.names_entrypoints
+
+
+def active(compiler="catalyst") -> bool:
+    """Check whether the caller is inside a :func:`~.qjit` evaluation context.
+
```

Comment:
I can imagine the design of this function being somewhat cumbersome. If I use a different compiler other than catalyst, I always have to specify it here as argument. Also, there might be functions which use this method without knowing which compiler the user is going to use.

A better design imo would be to return true if any available compiler is currently tracing.

Reply:
@dime10 I agree, although wonder if we should wait until/if we have multiple compilers available? because it's not clear to me if there is a nice way to do this right now, apart from just looping through every available compiler and checking for tracing 🤔 

 In the meantime, most developers would simply call this with no argument

Reply:
Right, but is there anything wrong with looping trough them? Seems like the way to go for me

Reply:
Yep looping would work. Alternatively, we could have a function that returns the name of the currently active compiler. E.g.,

```python
@qjit(compiler="catalyst")
def f(x):
    print(qml.compiler.active_compiler()) # will print "catalyst"
    return x ** 2
```

This would also solve your comment here: https://github.com/PennyLaneAI/pennylane/pull/4709#discussion_r1370877420 and https://github.com/PennyLaneAI/pennylane/pull/4698#discussion_r1370873444.

Now, functions like `while_loop` and `jacobian` _no longer need to have_ a `compiler` argument. Instead, they can work like this:

```python
def while_loop(...):
    active_compiler = qml.compiler.active_compiler()

    if active_compiler is not None:
        # call while_loop function of specified compiler

    # else, fallback to Python
    # (e.g., the catalyst.while_loop python fallback if available)
    ...
```

Reply:
I like @josh146's idea more than looping around available compilers! I don't personally like looping around available compilers. It adds unnecessary ambiguity to the code and support in my opinion. 

For functions like jacobian, we want to give the users the ability to write custom implementation for each available compiler. So, I'd assume someone would extend the patcher in the following way,

``` python
def jacobian(fn):
  if compiler.active("catalyst"):
     # do this
  elif compiler.active("catalyst2"):
     # do something else
  # else:
  # PL logic
``` 
For `while_loop` and `for_loop` on the other hand, I was explicit about the name of compilers which with the recent updates this is not necessary and I can follow the same approach and remove it.

Reply:
Hmm, that is an interesting question. Should different compilers be allowed to have functions with different signatures for each such primitive? Different implementations sure, but different signatures?

It would require hardcoding each compiler into the wrapper of each primitive, which is a less than ideal design 🤔

Reply:
@dime10 @josh146 I can think of an alternative here by keeping the logic and implementation details of multi-compiler support **only** in the Catalyst Frontend. This will remove the boilerplate code from PL src code and simplify the integration. There's also no need to pass the `compiler` in any of the compiler methods except `qjit` (or `qnode` depending on the scope of the multi-compiler support.. I also am in favor of passing it to `qjit` but we can discuss pros & cons later).

Reply:
@dime10 @josh146 let me know if you are okay with this update [9224e9b](https://github.com/PennyLaneAI/pennylane/pull/4692/commits/9224e9b3fb7c912b97134cce6107ad1f5237a71f)